File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        store = 0
        head = None
        ptr = head
        while (l1 is not None or l2 is not None):
            if (head is None):
                head = ListNode()
                ptr = head
            else:
                ptr.next = ListNode()
                ptr = ptr.next
            if (l1 is None):
                nodesum = l2.val + carry
                store = nodesum % 10
                carry = (nodesum - store)/10
                l2 = l2.next
            elif (l2 is None):
                nodesum = l1.val + carry
                store = nodesum % 10
                carry = (nodesum - store)/10
                l1 = l1.next
            else:
                nodesum = l1.val + l2.val + carry
                store = nodesum % 10
                carry = (nodesum - store)/10
                l1 = l1.next
                l2 = l2.next
            ptr.val = store
        if (carry == 1):
            ptr.next = ListNode()
            ptr = ptr.next
            ptr.val = carry
        
        return head

File: 2-add-two-numbers/add-two-numbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode* head = malloc(sizeof(struct ListNode));
    if (head == NULL) {
        return NULL;
    }
    struct ListNode* tail = head;
    int carry = 0;

    while (l1 != NULL || l2 != NULL || carry != 0) {
        int d1 = (l1 != NULL) ? l1->val : 0;
        int d2 = (l2 != NULL) ? l2->val : 0;

        int nodesum = d1 + d2 + carry;
        int store = nodesum % 10;
        carry = (nodesum - store) / 10;

        struct ListNode* newNode = malloc(sizeof(struct ListNode));
        if (newNode == NULL) {
            return NULL;
        }

        newNode->val = store;
        newNode->next = NULL;
        tail->next = newNode;
        tail = tail->next;

        l1 = (l1 != NULL) ? l1->next : l1;
        l2 = (l2 != NULL) ? l2->next : l2;
    }

    struct ListNode* toss = head;
    head = head->next;
    free(toss);

    return head;
}